src/Pos/Block/Network/Retrieval.hs

```diff
@@ -57,68 +57,103 @@ retrievalWorker getPeers = worker outs (retrievalWorkerImpl getPeers)
                 (Nothing, Nothing) -> retry
-                (Just v, _)        -> pure $ Left v
-                (Nothing, Just r)  -> pure $ Right r
-        either onLeft onRight loopCont
+                (Just chunk, _)    -> pure (handleBlockRetrieval chunk)
+                (_, Just rec')     -> pure (handleHeadersRecovery rec')
+        thingToDoNext
         mainLoop
-    onLeft ph =
-        handleAll (handleBlockRetrievalE ph) $
-        reportingFatal getPeers version $ workerHandle getPeers sendActions ph
-    onRight (peerId, rHeader) = do
-        logDebug "Queue is empty, we're in recovery mode -> querying more"
-        whenJustM (mkHeadersRequest (Just $ headerHash rHeader)) $ \mghNext ->
-            handleAll (handleHeadersRecoveryE peerId) $
-            reportingFatal getPeers version $
-            reifyMsgLimit (Proxy @(MsgHeaders ssc)) $ \limPx ->
-            withConnectionTo sendActions peerId $ \_peerData ->
-                requestHeaders mghNext peerId (Just rHeader) limPx
+    mainLoopE e = do
+        logError $ sformat ("retrievalWorker: error caught "%shown) e
+        throw e
+    --
+    handleBlockRetrieval chunk =
+        handleAll (handleBlockRetrievalE chunk) $
+        reportingFatal getPeers version $
+        workerHandle getPeers sendActions chunk
     handleBlockRetrievalE (peerId, headers) e = do
         logWarning $ sformat
             ("Error handling peerId="%build%", headers="%listJson%": "%shown)
             peerId (fmap headerHash headers) e
         dropUpdateHeader
         dropRecoveryHeaderAndRepeat getPeers sendActions peerId
+    --
+    handleHeadersRecovery (peerId, rHeader) = do
+        logDebug "Block retrieval queue is empty and we're in recovery mode,\
+                 \ so we will request more headers"
+        whenJustM (mkHeadersRequest (Just $ headerHash rHeader)) $ \mghNext ->
+            handleAll (handleHeadersRecoveryE peerId) $
+            reportingFatal getPeers version $
+            reifyMsgLimit (Proxy @(MsgHeaders ssc)) $ \limPx ->
+            withConnectionTo sendActions peerId $ \_peerData ->
+                requestHeaders mghNext peerId (Just rHeader) limPx
     handleHeadersRecoveryE peerId e = do
         logWarning $ sformat
             ("Failed while trying to get more headers "%
              "for recovery from peerId="%build%", error: "%shown)
             peerId e
         dropUpdateHeader
         dropRecoveryHeaderAndRepeat getPeers sendActions peerId
-    handleTop e = do
-        logError $ sformat ("retrievalWorker: error caught "%shown) e
-        throw e
 
 dropUpdateHeader :: WorkMode ssc m => m ()
 dropUpdateHeader = do
     progressHeaderVar <- ncProgressHeader <$> getNodeContext
     void $ atomically $ tryTakeTMVar progressHeaderVar
 
-dropRecoveryHeader :: WorkMode ssc m => NodeId -> m Bool
+dropRecoveryHeader
+    :: WorkMode ssc m
+    => NodeId
+    -> m Bool           -- ^ Whether given peer was kicked and recovery was
```

Comment:
Move this comment to function comment instead. I guess it's also worth mentioning the reason `peerId` should be passed. 
Consider scenario when you work with peer P and header H on recovery mode, it fails, at the same time other thread (new block arrived) replaces header to H2 and peer to P2, we drop communication with H2 and halt (while in fact we should ignore this exception and continue to work with P2).